src/ucp/api/ucp.h

```diff
@@ -107,6 +108,50 @@ BEGIN_C_DECLS
  */
 
 
+/**
+ * For internal use only.
+ */
+extern size_t ucp_request_internal_size;
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief A helper macro to allocate request on stack.
+ *
+ * A helper macro to allocate request on stack. In this case,
+ * @ref UCP_REQUEST_FLAG_EXTERNAL should be added to @ref ucp_request_t::flags
+ * initialization to hint UCP library.
+ *
+ * @note Useful for implementation of blocking operations like MPI_Send/MPI_Recv.
+ *
+ * @param [in]  _request_size   User request size, e.g. equal to
+ *                              @ref ucp_context_attr_t::requst_size.
+ *
+ * @return Pointer to ucp_request_attr_t for initialization.
+ */
+#define UCP_REQUEST_ALLOCA(_request_size) \
+    (ucp_request_attr_t *)((char *)alloca((_request_size) + \
+                                          ucp_request_internal_size) + \
+                           (_request_size) + ucp_request_internal_size - \
+                           sizeof(ucp_request_attr_t));
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief A helper macro to get access to context associated with @a _request.
+ *
+ * A helper macro to get access to context associated with @a _request returned
+ * by ucp_<tag|stream>_<send|recv>_nbx function as return value or passed to
+ * completion callback.
+ * 
+ * @param [in]  _request       Request handle.
+ *
+ * @return Pointer to user context.
+ */
+#define UCP_REQUEST_CONTEXT(_request) \
+    ((ucp_request_t *)(_request) - 1)->context;
```

Comment:
how can ucp_request_t type be sued in API?
i'd suggest to add at least smoke test which uses the macros UCP_REQUEST_ALLOCA and UCP_REQUEST_CONTEXT